### PR TITLE
Release of dqm-ml v2.0.0-rc

### LINKED ISSUE #11: Doc generation with mkdocs
As a developer, I want to generate documentation in html format from the readme

1. Choose a mkdocs theme, for instance:

- readthedocs
- mkdocs
- material

2. Configure mkdocs (mkdocs.yml)

3. Write/update hooks in docs/hooks.py to:
- Copy the readme and notebooks to the docs folder
- Update relative links in the readme

### LINKED ISSUE #22: Crash when configured output folder does not exists (useless loss of time for user)
At the end of the computation, if the path for output parquet file is (defined in the config file points) to a directory that does not exist, an error is raised and the Parquet file is not written, which is unfortunate because user have spent time for computations, and results are in memory but are lost because output cannot be created.

2 possible solutions 

- the output directory should be created on the fly, 
- Or the existence of the output directory existence should be checked **before** running the computation to avoid user loosing time.

### LINKED ISSUE #24: Add some message about progression during the post_processing part of computation
The whole processing of a pipeline configuration seems to be spolitted into two parts : 

- Loop on all batch
-  Post processing :

For the l1st part, the loop on batch, a progress bar has been added that is great for users to follow the progression.

But during the second part, nothins is displaying, and this part can spend a nun-negligible time .

Thus I suggest to add at least a message,, with something like: "post processing, is runnning, .. ." and eventually a progress bar if it is possible

### LINKED ISSUE #25: Enable computation from multiple data selection in the same job

### LINKED ISSUE #26: Enable to compute domain gap between more than 2 data selection

### LINKED ISSUE #27: Enable to configure data loader to parse dataset for multiple dataselection

### LINKED ISSUE #28: Move tests from each package to the workspace

### LINKED ISSUE #29: Rename dqm-ml-pipeline in dqm-ml-job

### LINKED ISSUE #31: Need upgrade in test package dependency
after uv sync --upgrade 

We have the following errors : 

ImportError while loading conftest '/home/loic/work/Safenai/opensource-repos/dqm-ml/tests/conftest.py'.
tests/conftest.py:5: in <module>
    import fiftyone.zoo as foz
.nox/test-3-12/lib/python3.12/site-packages/fiftyone/__init__.py:31: in <module>
    from fiftyone.__public__ import *
.nox/test-3-12/lib/python3.12/site-packages/fiftyone/__public__.py:244: in <module>
    from .core.session import (
.nox/test-3-12/lib/python3.12/site-packages/fiftyone/core/session/__init__.py:10: in <module>
    from .session import close_app, launch_app, Session
.nox/test-3-12/lib/python3.12/site-packages/fiftyone/core/session/session.py:37: in <module>
    from fiftyone.core.state import build_color_scheme, StateDescription
.nox/test-3-12/lib/python3.12/site-packages/fiftyone/core/state.py:28: in <module>
    from fiftyone.server.scalars import JSON
.nox/test-3-12/lib/python3.12/site-packages/fiftyone/server/scalars.py:19: in <module>
    BSON = gql.scalar(
.nox/test-3-12/lib/python3.12/site-packages/strawberry/types/scalar.py:295: in scalar
    return wrap(cls)
           ^^^^^^^^^
.nox/test-3-12/lib/python3.12/site-packages/strawberry/types/scalar.py:274: in wrap
    warnings.warn(
E   DeprecationWarning: Passing a class to strawberry.scalar() is deprecated. Use StrawberryConfig.scalar_map instead for better type checking support. See: https://strawberry.rocks/docs/types/scalars
nox > Command pytest --cov=packages/dqm-ml-pipeline/src --cov=packages/dqm-ml-core/src --cov=packages/dqm-ml-pytorch/src --cov=packages/dqm-ml-images/src --cov=packages/dqm-ml-v2/src --cov-report=html:docs/static/htmlcov --cov-report=term --cov-fail-under=1 --html=docs/static/pytest_report.html tests/unit tests/integration tests/cli failed with exit code 4
nox > Session test-3.12 failed.

### LINKED ISSUE #41: Take into account security warning

### PR DESCRIPTION
**Pull request by lcantat**
## 2.0.0-rc (2026-04-10) - Release Candidate for V2

This release is the V2.0.0 release candidate, featuring major refactoring, security fixes, and documentation improvements.

### Issue ticket number and link

* fix: #61 - Python 3.10/3.11 compatibility
* fix: #60 - Fix parquet dataloader with image_path column
* fix: #59 - Data selection tests for CLI
* fix: #41 - Security warnings addressed
* fix: #31 - Test package dependencies upgraded
* fix: #29 - dqm-ml-pipeline renamed to dqm-ml-job
* fix: #28 - Tests moved to workspace root
* fix: #27 - Multi data selection configuration
* fix: #26 - Domain gap computation for multiple selections
* fix: #25 - Multiple data selection in same job
* fix: #24 - Progress messages during post-processing
* fix: #22 - Output folder auto-creation
* fix: #11 - MkDocs documentation generation

### Breaking Changes

* **Package rename**: `dqm-ml-v2` → `dqm-ml` (CLI entry point)
* **CLI command**: `dqm-ml-v2` → `dqm-ml`
* **Legacy removal**: Removed dependency on legacy `dqm-ml` submodule

### Security Fixes (10 vulnerabilities fixed)

| Package | Old Version | New Version |
|---------|-------------|-------------|
| strawberry-graphql | 0.287.3 | 0.314.3 |
| tornado | 6.5.4 | 6.5.5 |
| cryptography | 46.0.4 | 46.0.7 |
| requests | - | 2.33.0 |
| pygments | - | 2.20.0 |

CVEs addressed: CVE-2026-35526, CVE-2026-35523, CVE-2026-31958, CVE-2025-47287, CVE-2026-26007, CVE-2026-34073, CVE-2026-37990, CVE-2026-25645

### New Features

* **Test Strategy Documentation**: Comprehensive testing docs in `contributing.md` with mermaid diagrams
* **Git LFS Guide**: Added to `quickstart.md` for development setup
* **Test Fixtures**: Session-scoped fixtures (`coco_data`, `normal_dist`, `uniform_dist`, etc.)
* **Test Utils**: New `tests/utils/` with helpers for files, jobs, plots
* **tqdm dependency added**: Added to `dqm-ml-job` for progress display

### Documentation Improvements

* **AGENTS.md**: New contribution guidelines for AI agents
* **Test Strategy**: Complete testing documentation with diagrams
* **Git LFS Section**: Added to quickstart for large file handling
* **TODO/ROADMAP Sync**: Aligned V1 metrics lists and version references
* **Mermaid Diagrams**: Fixed dependency direction (core ← job, not →)
* **All docs updated**: Replaced `dqm-ml-v2` references with `dqm-ml`
* **New CLI Reference**: `docs/cli.md`
* **New Data Loaders**: `docs/dataloaders.md`
* **New Python Compatibility**: `docs/python_compatibility.md`
* **Navigation**: Organized into separate pages in mkdocs.yml

### Technical Changes

* Pinned `mkdocs-jupyter<0.26` to fix api-autonav compatibility
* Refactored `conftest.py` - restructured fixtures
* Added deprecation warning filters for strawberry and fiftyone

### New CLI and Integration Tests

* Test domain gap matrix between class pairs
* Test quickstart examples with completeness config
* New test fixtures and config files in `tests/fixtures/getting_started/`
* Test data: COCO dataset with class information (102+ classes)
* 18 tests added (unit + CLI)

### Python Compatibility (Issue #61)

* **Python 3.10/3.11 support added**
* Fixed `typing.override` import: Changed from `typing` to `typing_extensions` in 9 package files
* Fixed fiftyone lazy import in test fixtures (glob2 SyntaxError workaround)
* CI now tests: Python 3.10, 3.11, 3.12, 3.13
* New documentation: `docs/python_compatibility.md`

### Data Selection Features (Issues #25, #27, #59, #60)

* **split_by configuration**: Create multiple selections from single dataloader based on column values
* **split_values**: Specify values to create selections (e.g., split_by: class, split_values: [dog, cat])
* **filter configuration**: Filter rows by column values
* New fixtures: `coco_classes`, `bird`, `elephant` for domain gap testing
* Example configs: `domain_gap_split_2classes.yaml`, `domain_gap_split_top10.yaml`, `domain_gap_matrix.yaml`

### Dependencies Updated

* fiftyone: 1.14.1 → 1.13.0
* torch: 2.10.0 → 2.11.0
* torchvision: 0.25.0 → 0.26.0
* Many other transitive dependencies upgraded

### Testing

* 72 tests passing (18 new tests)
* Integration test fixtures for real data testing
* CLI end-to-end tests
* Test coverage reports available in `docs/reports/`
* Python compatibility tests: 3.10, 3.11, 3.12, 3.13

### CODE REVIEW

**Comment by Josk187** [comment]
I'm referring to branch "dev" behind the main pull request /

Two first questions:

- In the main  readme file, badges refers to python 3.12 and 3.13, whereas in pkg toml file we have--> python >= 3.10
 Thus , is the compatibility bounded to 3.12 and 3.13 only, ?? which information is correct ? readme or toml ?

- "Diversity" metric was present in dqm-ml_v1, it seems that this metrics has been removed in dqm-ml v2, (or I missed something ? ) that makes a little functional regression from previous version,   is it planned to put it back in a next version ??? 

- Requirements in all subpackages are given with inequality such "pkg_name">= X.X  that make the installed code not fully controlled. uv.lock contains a snapshot of a workable config but requires to use "uv" to be used and cannot be used classical by pip install dqm-ml from pypi. Moreover uv.lock contains  all dependencies for all packages whereas the followed philosophy seems to have an uv.lock for each subpackages ( see with @yann-gaste if this is ok )

- Slight comment-->  pyarrow>=6.0.0  dependancy  is present in dqm-pytorch  but it is already present in dqm-ml-core 

Thanks

**Comment by lcantat** [comment]
> I'm referring to branch "dev" behind the main pull request /
> 
> Two first questions:
> 
> * In the main  readme file, badges refers to python 3.12 and 3.13, whereas in pkg toml file we have--> python >= 3.10
>   Thus , is the compatibility bounded to 3.12 and 3.13 only, ?? which information is correct ? readme or toml ?
=> I will correct, thanks for the remark
=> You are right, I change to add check for 3.10 and 3.11 compatibility (we just do not test several case which are not structural but data linked so not subject to compatibility. I add also a dedicated section for compatibility.

> * "Diversity" metric was present in dqm-ml_v1, it seems that this metrics has been removed in dqm-ml v2, (or I missed something ? ) that makes a little functional regression from previous version,   is it planned to put it back in a next version ???
==> It will be integrated before 2.0.0 will be release, it is mark as a limitation of the 2.0.0_rc. There implementation need discussion with faouzi and other to validate interface proposed. (confusion between metric in representativness and diversity)

> * Requirements in all subpackages are given with inequality such "pkg_name">= X.X  that make the installed code not fully controlled. uv.lock contains a snapshot of a workable config but requires to use "uv" to be used and cannot be used classical by pip install dqm-ml from pypi. Moreover uv.lock contains  all dependencies for all packages whereas the followed philosophy seems to have an uv.lock for each subpackages ( see with @yann-gaste if this is ok )

==> We can discuss this point if needed and publish a referenced used version of packages version that where used during tests (the one in uv.lock. If it's a good practice for a project to pin version used to prevent regression, it'is not the same for a library that we want to be used by a large number of project. We can not export strict constraint. (see https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/#install-requires) :
**_It is not considered best practice to use install_requires to pin dependencies to specific versions, or to specify sub-dependencies (i.e. dependencies of your dependencies). This is overly-restrictive, and prevents the user from gaining the benefit of dependency upgrades._**
 

> * Slight comment-->  pyarrow>=6.0.0  dependancy  is present in dqm-pytorch  but it is already present in dqm-ml-core

Both need the dependency, as it is in core we can remove in pytorch, be maybe in futur one will need a bigger version and not the other. To be discuss

> Thanks

**Comment by Josk187** [comment]
Thanks for answer, @lcantat  

Depending of @yann-gaste  point of view, If needed a solution about bounded requirements version question, to be not uselessly restrictive for common users,  one possible  solution could be to add a specific extras dqm-ml[locked]  (eventually many for different python version )  for user that want in a specific context (for example in a production docker container) install directly a guarranted safe configuration. 

Another observation, in readme.md file 👍 

Link to jupyter notebookl is wrong, and need to be updated .
screenshot

**Comment by Josk187** [comment]
In  example/scripts/completeness.py -->replace deprecated

exec_qml_job(config["pipeline_config"]) 

by  

exec_qml_job(config["config"])

**Comment by lcantat** [comment]
I merge this version into main, this will be the 2.0.0-rc tag.
